### PR TITLE
deprecate `typealias` in favor of `const` assignment or `A{B} = ...`

### PR DESCRIPTION
**Pull request by JeffBezanson**
`typealias` is now completely redundant, as it is just a `const` assignment of a name to a type. I also added the syntax

```
Vector{T} = Array{T,1}
```

in analogy to short-form function definitions.

### CODE REVIEW

**Comment by KristofferC** [comment]
When do you have to use `const` and when can the short form be used? And if you forget the `const` will it just give you bad performance?

File: base/array.jl [hunk 1]
```diff
@@ -4,22 +4,22 @@
...
-typealias NeedsShaping Union{Tuple{Integer,Vararg{Integer}}, Tuple{OneTo,Vararg{OneTo}}}
-
-typealias Vector{T} Array{T,1}
-typealias Matrix{T} Array{T,2}
-typealias VecOrMat{T} Union{Vector{T}, Matrix{T}}
-
-typealias DenseVector{T} DenseArray{T,1}
-typealias DenseMatrix{T} DenseArray{T,2}
-typealias DenseVecOrMat{T} Union{DenseVector{T}, DenseMatrix{T}}
+AbstractVector{T} = AbstractArray{T,1}
```
Comment (yuyichao) [inline base/array.jl]: Why do these not have const?

**Comment by JeffBezanson** [comment]
In function definitions `f(x) = x`, `const f` is inserted, and the same happens here. `a{b} = c` is also a `const` assignment. These are usually global variables, so the same performance issues apply. However these are very often used only in method signatures, where they're evaluated once at the top level, in which case there's no performance impact.

**Comment by JeffBezanson** [comment]
I just realized `X{} = ...` (the zero parameter case) works, and since it needs no `where` expressions, is equivalent to just a `const` assignment. Not sure if feature or bug.

**Comment by stevengj** [comment]
I'm all for getting rid of `typealias`.

However, I'm skeptical that we want a new abbreviated syntax here.  Parameterized typealiases are a lot less common than one-line functions, and every new syntax we add also makes the language more complicated to learn.  `const Vector = Array{T,1} where T` isn't too bad.

**Comment by JeffBezanson** [comment]
I'm fine with allowing `const Vector{T} = Array{T,1}`. But without `const`, it will still be constant :)

**Comment by JeffBezanson** [comment]
Ah, fortunately due to sufficiently generic code `const A{B} = C` already works, as does `const f(x) = x`. I think I will disallow `A{} = B` though since it's not meaningful.

**Comment by yurivish** [comment]
What are the practical differences between saying `Vector = Array{T,1} where T` (without `const`) and `const Vector = Array{T,1} where T`?

**Comment by StefanKarpinski** [comment]
If `Vector` is non-const then anything that references it will have bad performance. Similar to using non-const globals in general.

**Comment by JeffBezanson** [comment]
Yes, it's just a normal const vs. non-const binding.

**Comment by yurivish** [comment]
Thanks. I like the idea of a convenient syntax for the parametric case but am concerned that the performance subtleties will continually trip people up if the const is implicit. 

I'm pretty familiar with Julia, but the parallels between the new syntax (`A{T} = B{T, 1}`) and normal variable declaration (`A = B`) feel like they'd pull me strongly towards accidentally leaving a `const` off of a typealias of the latter form. 

Julia itself has [quite a few](https://github.com/JuliaLang/julia/search?utf8=%E2%9C%93&q=typealias) typealiases without parameters, so it seems to be a pretty common thing.

**Comment by tkelman** [comment]
looks like this test https://github.com/JuliaLang/julia/blob/6d6a3c3c20a38f02c9fe3c3bec3b40685c91cb27/test/parse.jl#L701 now needs extra spaces around the `&&` ? seems kinda unrelated but harmless enough?

**Comment by JeffBezanson** [comment]
> am a little worried that the performance subtleties will trip people up if the const is implicit.

Agreed; it would really be nice if global bindings could be const by default (and use e.g. `Ref` for mutating). That would be a good solution everywhere but the REPL it seems.

**Comment by StefanKarpinski** [comment]
> it would really be nice if global bindings could be const by default ... That would be a good solution everywhere but the REPL it seems.

What about everything being `const` once the module is closed? "How would this help a script in a file?" you ask – since the whole script runs before the `Main` module is closed. However, if `Main` has a `main()` function we could automatically called when the script ends – after the `Main` module is closed, so the constness of bindings would actually be reliable at that point.

**Comment by JeffBezanson** [comment]
That could work. Of course if any code gets compiled before the module closes, it would not be able to take advantage of const-ness. Maybe it could be a per-module property; you can say that a certain module should have non-const-by-default bindings, which would normally only be used by the repl.

**Comment by StefanKarpinski** [comment]
We could keep track of methods that were compiled before the module was closed and which depended on a global. Closing the module would then invalidate their code, forcing them to be recompiled.

**Comment by JeffBezanson** [comment]
Yes we could do that, but compiling things twice is going to hurt load times even more.

**Comment by JeffBezanson** [comment]
Ok, I took out the change to `deparse` that broke a test since it doesn't belong here. The extra spaces do look better, but that can be done separately.

**Comment by StefanKarpinski** [comment]
If a function uses a global while its module is still open, that would be a good heuristic to decide to use an interpreter – if we end up having a full-fledged one. That way compilation would occur:

1. On first usage if the function only uses const globals;
2. On first usage after the module is closed otherwise.

That way the compilation time wouldn't be affected.

**Comment by iamed2** [comment]
The behaviour with `const A{B} = C` is kind of confusing for me:
```julia
julia> const Foo{Int} = Vector
Array{T,1} where T

julia> Foo
Array{T,1} where T

julia> Foo{Int}
Array{Int64,1}
```
```julia
julia> const Bar{Int} = Float64
Float64

julia> Bar
Float64

julia> Bar{Int}
ERROR: TypeError: Type{...} expression: expected UnionAll, got Type{Float64}
```
```julia
julia> const Baz{1} = Int
ERROR: syntax: malformed expression
```

Maybe these could use better errors?

**Comment by JeffBezanson** [comment]
I'll fix the `malformed expression` one.

The `B` in `A{B} = C` is a formal argument name, like `f(Int) = Int`. I know what you mean, but calling an argument `Int` does obscure things a bit.

**Comment by iamed2** [comment]
Ah thanks, the first example makes sense now.

For the second, is the extra argument being ignored in the definition but not in usage? I would expect that if `Bar{Int}` is going to error then `const Bar{T} = Float64` should error.

**Pull request merged**